app/graphql/types/query_type.rb

```diff
@@ -0,0 +1,12 @@
+Types::QueryType = GraphQL::ObjectType.define do
+  name 'Query'
+  description 'The query root of this schema.'
+
+  field :search, !types[Types::OrganizationType] do
```

Comment:
The semantics here is a little confusing - 'search' is a `field` on a `QueryType` root, but I imagine adding an endpoint to fetch Organizations by ID would be an 'organization' field in the same query root. That sounds strange, as we end up mixing entities with arbitrary operations on a flat query namespace. Is there some best practice for how to structure that better? Can they be nested for example?

I'm also wondering what the `!` signifies here - we don't use it in our declarations in GravQL (like [here](https://github.com/cavvia/gravity/blob/master/lib/graph/queries.rb#L7)).

Reply:
The `!` means "required", but in this context I believe it does nothing, so removed.

I am with you wrt API query schema. This PR wanted to refactor the existing API "as is", so lets try to do what you suggest after it's merged? And I am a GraphQL noob so maybe @craigspaeth can help us out with what this should be?

Reply:
The `!` means non-null, you should definitely include it _everywhere_ where null will never be returned. I.e. does this return null or an empty list when no matches are found?

Reply:
I'm not caught up on the latest thinking in schema design, but the nice thing about having a schema that can evolve in an "append only" manner with auto-docs and not risking bloating response size, is you have more leeway to mix "RPC" and "resourceful" styles.

One way to think about designing GraphQL is in a resourceful manner which you can carry over your REST/CRUD design principles—so what would be POST/GET/PUT/DELETE would be...

```
query organization(id: 123) {}
query organizations(term: "david") {}
mutation createOrganization(name: "foo") {}
mutation updateOrganization(id: 123 name: "foo") {}
mutation deleteOrganization(id: 123 name: "foo") {}
```

In this case for search what would be `http://api.com/api/organizations?term="david"` in REST would be `organizations(term: "david")` in GraphQL.

That said, any style is fair game as GraphQL doesn't prescribe a style itself, but the community is [starting](https://blog.graph.cool/designing-powerful-apis-with-graphql-query-parameters-8c44a04658a9) to [talk](https://dev-blog.apollodata.com/designing-graphql-mutations-e09de826ed970) [about](https://www.youtube.com/watch?v=R8Ny0i5S978) [best practices](https://medium.com/impraise-design-engineering/making-mistakes-with-graphql-874b8ca62b9d).

Reply:
@alloy So this returns an empty array when nothing is found. Is that the same thing in GraphQL context? Does this mean it still needs a `!`? Where is the doc for this, I couldn't find it.

Reply:
I am with @craigspaeth on the resourceful way of designing this. I'd like to do this in a subsequent PR and kill the search API. Again my goal for *this* PR was not to redesign the API but change the layout to be GraphQL.

Reply:
Yeah, then you need to add it. There’s no specific documentation for this regarding collections (afaik), the important part is that in the schema `organizations` is just a field that returns something, it could be null or non-null like any other fields.

Reply:
Not sure if you were only addressing @cavvia’s initial comment, but as far as making this a good example goes, you should really add the `!` back imo. The schema is now harder to use because it’s telling the clients that they need to perform a null check before using the results, which isn’t necessary.

Reply:
Added.

Reply:
🙏🏽

Reply:
I think @craigspaeth's resourceful approach (`organizations(term: 'david')`) is better than what we have now and is probably a more conventional schema design. If a GraphQL query is about selecting fields on objects, rather than simply nesting agnostic 'nodes', it makes sense to use it in a more resourceful manner, rather than mixing an RPC style in there with verbs such as 'search'. Having said that, mutates essentially look RPC in style! So it's a mixed bag.

@dblock Sure we can follow up with schema improvements in further PRs, as there's no live client yet, so we have that luxury.